[1/3] Restructure LIBDIR: Move Dynlink, Str and Unix to sub-directories

dra27 opened the pull request:
This PR is the first and main PR of a brave (in the Apple sense) series of changes which seek to separate the components in OCaml's `$LIBDIR`. This PR specifically moves the Dynlink, Str and Unix libraries into sub-directories of `$LIBDIR`. The idea of the change was provisionally agreed at the last developers' meeting.

The motivation for this change is hygiene in build systems: at present, it is very difficult to impossible for Dune (for example) to ensure that a library specifies its dependency on the Unix library.

The change itself is quite simple, and in the first commit. A small tweak to two `Makefile`s ensures the libraries are installed to their new locations. Any previous artefacts are removed (cf. #1724, #1728 and  #10301, noting that the removal of artefacts is "precise" w.r.t. basenames).

That change on its own is very breaking, of course. 

The rest of the commits relate to backwards compatibility. The compilers, toplevel and debugger will all automatically add the required include path to the load path, but display an alert for each directory as this happens. For example:

```
$ ocaml unix.cma
File "_none_", line 1:
Alert ocaml_auto_include: OCaml's lib directory layout changed in 5.0. The unix
subdirectory has been automatically added to the search path, but you should add
-I +unix to the command-line to silence this alert.
OCaml version 5.0.0+dev0-2021-11-05
Enter #help;; for help.
```

or similarly with:

```
$ ocaml
OCaml version 5.0.0+dev0-2021-11-05
Enter #help;; for help.

# type t = Unix.file_descr;;
Alert ocaml_auto_include: OCaml's lib directory layout changed in 5.0. The unix
subdirectory has been automatically added to the search path, but you should add
-I +unix to the command-line to silence this alert.
type t = Unix.file_descr
```

In implementing this alert, I've attempted to avoid penalising code which has specified the appropriate include directories, especially as we know that the load path is a relatively "hot" part of the build for some users. The mechanism I've gone for is to add an extra attempt around `Load_path.find` and `Load_path.find_uncap` for the case when there is an actual miss on the load path. In almost all cases, a miss on the load path is fatal. The exceptions are `#remove_directory` in the toplevel and gaining inlining information from `.cmx` files when linking in `ocamlopt`, so at less critical stages. The toplevel, compilers and debugger add `"+dynlink"`, `"+str"` and `"+unix"` to a "secondary" load path which is lazily scanned on the first miss on the normal load path. If the file being searched (say `unix.cma` and `unix.cmi`) is found, then that directory's entries are promoted to the main load path, and the fact that this took place is recorded. From the user's perspective in, say, the toplevel what happens at `#load "unix.cma"` is as if the user issued `#directory "+unix"` and then tried the `#load` again.

`Load_path` is very low-down the dependency graph of compiler-libs, so rather than displaying alerts itself, it simply maintains a list of directories which have been automatically added. This list is queried by functions added with the other alert functions in, ahem, `Location`.

The initialisation code is a duplicated between the debugger, driver and toplevel. It's not huge, but I'm not sure exactly where to put a function to do that - how averse are we to adding modules to ocamlcommon?

Another option might be to install the files in both places, but this isn't a silver bullet either - including the new directory can cause warnings in some tools (`ocamlfind` doesn't like finding `.cmi` files in multiple directories, for one) and it seems to be harder to record the need to display an alert without altering main load path data structure, which is a much more intimidating prospect than altering an error path.

Very similar work was previously attempted and abandoned in #1569. This work differs both in implementation and timing in a few ways which hopefully means we can reach a consensus on merging it this time. The key differences:
- OCaml 5.0 already breaks quite a few older packages, which means that this change is not as sensitive as it was in 2018.
- We have superior testing systems in place to assess the impact of this change. For OCaml 5, accepting that both ocamlfind and dune require patching, only **one** other package is broken in the whole of opam-repository by this PR, and that's an old release of that particular package, the current release using Dune and being unaffected.
- No changes are made to compiler-libs.
- Rather than implementing new compiler flags, this PR displays an alert

I have tested this change quite extensively on 4.14 - in many cases, the packages affected are already lying to their build systems. Dune and ocamlbuild users, for example, should not use `Unix` without having `unix` in their `libraries` stanza or `use_unix` in their `_tags` file. Even packages building directly with `ocamlfind` should include `-package unix` in the command-line. These packages can be fixed regardless of whether this PR is merged. Given a flurry of PRs now with this change (these packages can merge these updates regardless of whether this PR is accepted), I imagine that by OCaml 5.2, the number of packages still displaying this alert will be close to zero and the entire mechanism can then be reverted - at which point this change lives entirely in the build system.

There are other packages which contain commands of the form `ocaml unix.cma` which want to become `ocaml -I +unix unix.cma` - PRs are opened for these as well.

There are three possible pieces of future work which I haven't done yet, mainly because they are more philosophical (separating the otherlibs out mitigates actual issues):
1. C stubs for otherlibs are still installed to `$LIBDIR` (this is already the case for systhreads, so this is partially done to stay consistent). This could be trivially fixed for all these, but it could be viewed as separate. Note that at present we're saying that `-L /usr/lib/ocaml` is all you need to give to link _in C_. I can't remember the reference, but there's at least one application in the wild which links manually with our C stub libraries, because I remember dealing with a bug report relating to it...
2. `std_exit.cmo` and `std_exit.cmx` remain, despite the fact these are really part of the compiler, and not the Standard Library. The compilers impose a requirement that a standard library replacement implements a function called do_at_exit, but the actual implementation of std_exit is part of the compiler, not the stdlib. Indeed, in many ways there's no need for a module called `Std_exit` at all, we could generate a temporary file.
3. The runtime and stdlib still share the same directory. This is also a more principled matter, than actually causing issue: there's an open question for whether the runtime should go in `+runtime` or the stdlib in `+stdlib`. FWIW, I would advocate the latter - `lib/ocaml` now reflects the files of the OCaml _runtime_ and _compiler_ with the sub-directories all being libraries. But that can readily be for another day and another release...

dra27 committed:
Install otherlibs to separate directories

dra27 committed:
Add Warnings.ghost_loc_in_file

Factors out some common usage between Warnings and Location for
zero-length ghost locations.

dra27 committed:
Add a hook for misses in Load_path

Load_path.find{,_uncap} now invoke a hook before raising Not_found,
which provides a mechanism for automatically adding directories to the
Load_path if required.

dra27 committed:
Display an alert if -I +unix, etc is missing

dra27 commented:
(temporarily adding the no-change-entry-needed label just to clear the hygiene check for now)

nojb reviewed (commented):
Thanks for this PR @dra27!

Just playing devil's advocate: I'm wondering if we really need the backwards compatibility logic at all? After your associated PRs are merged upstream, the amount of breakage will be rather minor, isn't it? (Anyone using `ocamlfind` or `dune` will be automatically "patched".) So perhaps we could just do the breaking change and save on the added complexity introduced by the backwards compatibility story?

On the other hand, you already did the implementation work, so maybe not worth spending too much time on this question.

dbuenzli commented:
This is very good. Thanks @dra27 for getting into these unglamorous bits. Here are a few comments:

1. Regarding:
     
   > 1\. C stubs for otherlibs are still installed to `$LIBDIR` (this is already the case for systhreads, so this is partially done to stay consistent). This could be trivially fixed for all these, but it could be viewed as separate.

     I couldn't infer the actual reason for that. Is there any load path issue ? This makes OCaml distributed libraries still special "in some way" and prevents uniform treatment by build tools. I would really prefer if that wasn't the case.

3. I wonder a bit if the backward compatibility measure is not a bit overkill especially as it adds more complexity to the already quite convoluted OCaml path. Is it really worth it ? Maybe good upfront communication would be an equally good solution.

2.  > 3\. The runtime and stdlib still share the same directory. This is also a more principled matter, than actually causing issue: there's an open question for whether the runtime should go in `+runtime` or the stdlib in `+stdlib`.

    Not sure about the exclusiveness here. They could also both get their own directory.

dra27 commented:
Actually, @nojb, I was originally aiming for no backwards-compatibility story... the alert was added quite some way in. If we're happy with the breakages, I'd be more than happy to lose the complicated load path and alerting tweaks! The packages I've opened PRs for are ones I patched in order to open up more packages for testing on 4.14 - there are more. I'll kick off an opam-health-check run against 5.0 and see just how "bad" it is.

@dbuenzli - I also couldn't see an immediate reason for the separation of the stub libraries. Truth be told, I forgot about the stub libraries (they're prefixed...!) and was about to correct it when I noticed that systhreads has always done it that way! It's no issue for OCaml, as its invocations of the linker should get the same load path. I also agree with the idea of `lib/ocaml` consisting only of sub-directories.

v-gb commented on otherlibs/Makefile.otherlibs.common:
```diff
@@ -96,15 +98,21 @@ install::
+# from the previous installation.
+	rm -f $(addprefix "$(INSTALL_LIBDIR)"/, \
+          $(LIBNAME).cma $(CMIFILES) \
+          $(CMIFLES:.cmi=.mli) $(CMIFILES:.cmi=.cmti) \
```
```suggestion
          $(CMIFILES:.cmi=.mli) $(CMIFILES:.cmi=.cmti) \
```

With this, incremental installation should be correct. Currently you get this difference between installing on top of old installation vs installing from scratch:

```diff
$ diff -u <(cd /tmp/installed-base-then-tip/lib/ocaml; find | sort) <(cd /tmp/installed-tip/lib/ocaml; find | sort)
--- /dev/fd/63	2022-04-25 00:14:01.665105415 -0400
+++ /dev/fd/62	2022-04-25 00:14:01.665105415 -0400
@@ -1947,7 +1947,6 @@
 ./stringLabels.mli
 ./string.ml
 ./string.mli
-./str.mli
 ./str/str.a
 ./str/str.cma
 ./str/str.cmi
@@ -1983,8 +1982,6 @@
 ./unit.ml
 ./unit.mli
 ./unix
-./unixLabels.mli
-./unix.mli
 ./unix/unix.a
 ./unix/unix.cma
 ./unix/unix.cmi
```

dra27 replied:
Ouch - thanks!

v-gb commented on otherlibs/Makefile.otherlibs.common:
```diff
@@ -96,15 +98,21 @@ install::
 ifneq "$(STUBSLIB)" ""
 	$(INSTALL_DATA) $(STUBSLIB) "$(INSTALL_LIBDIR)/"
```
I know you didn't do this on purpose, but it seems strange to me to separate the .a for the C files from the .a of the cmxa. It's preferable that the whole library be together in one directory, the same way that every non compiler library in the rest of the ecosystem has to work.

```suggestion
	$(INSTALL_DATA) $(STUBSLIB) "$(INSTALL_LIBDIR_LIBNAME)/"
```

dra27 replied:
I agree - I think I may open PR 4/3 to cover this one, given that it affects systhreads too.

v-gb commented on utils/load_path.ml:
```diff
@@ -66,10 +77,11 @@ let prepend_add dir =
     ) dir.Dir.files
 
-let init l =
+let init ?(secondary_dirs=[]) l =
```
I'm wondering if this should be a required parameter. There are three calls to `Load_path.init` that don't pass this (in the debugger and ast_mapper.ml), and they look suspicious.

dra27 replied:
I had looked at those and decided they didn't/couldn't need it, but can stare again!

v-gb commented on debugger/main.ml:
```diff
@@ -229,7 +229,14 @@ let main () =
     Loadprinter.init();
-    Load_path.init !default_load_path;
+    let secondary_dirs =
+      if !Clflags.no_std_include then []
```
It seems like there should be `Clflags.std_secondary_include_dirs` with this logic.

dra27 replied:
Ah, that's a good idea - so no flag, but the fact it's predicated on `Clflags.no_std_include` does suggest a good home there, yes.

v-gb commented on debugger/main.ml:
```diff
@@ -229,7 +229,14 @@ let main () =
+        Location.initialise_auto_include_alerts expand
+          Config.automatically_included_otherlibs
+    in
+    Load_path.init ~secondary_dirs !default_load_path;
```
I wonder if it wouldn't be simpler overall to pass into `Load_path.init` a function that can report these new alerts. This way, we can be sure that the alert is always reported when needed, and the code wouldn't be split between load_path.ml, location.ml, and code sprinkled all over.

dra27 replied:
Yes, that's a very nice way of short-circuiting it, thanks!

v-gb reviewed (commented):
The change of installation dir is straightforward, but the backwards compability part is more involved. I'm not sure either how necessary the compatibility is (yes, it's a breaking change, but the alert only helps with the kind of breaking change that's fixed with a simple `-I +unix`, not with breakage of things like `ocamlopt $(ocamlc -where)/unix.cmxa`).

It we do want to keep it, maybe it's not terrible to do something more special purpose but simpler, like:

```
(* Load_path.find *)
let find fn =
  try ... the existing code
  with Not_found ->
    if is_basename fn && List.mem [ "dynlink"; "unix; "unixLabels"; ... ] (Filename.chop_ext (uncapitalize fn))
    then (
      let x = Misc.find_in_path [ "+unix"; "+str"; "+dynlink] fn in
      generate_alert ();
      x
    ) else raise Not_found
```

dra27 commented:
Thanks for the review, @sliquister - making the alert function a callback will definitely tidy-up that compatibility code considerably. I'll see what going totally "ad hoc" with `Load_path.find` and `Load_path.find_uncap` looks like - my initially assumption was that that would never pass review 😉

shindere commented:
Many thanks, @dra27, for this really cool PR.

I agree with everything said so far.

I didn't read any comment saying we should take so much care to
backward compatibility, neither in temrs of alerts, nor in temrs
of installing.

So I'd avoid as much complexity as possible. Can't we simply
ask people to install in a clean tree, rather than dealing with a
possibly unclean one?

My only comment about the code is regarding `"$(INSTALL_LIBDIR_LIBNAME)/"`.
Why insist onhaving the final `/`?

Would also love to have `/usr/lib/ocaml` to contain only subdirectories, and
having one for the runtime and one for the stdlib looks very sensible to me,
withthe C stubs in the
same directory than the library they belong to.

It feels like, once all this has been sorted out, it would be time to
revisite #11007.

dra27 commented:
Thanks for the various comments and reviews! I've just pushed a revised version - still with the alert, for now. Addressing various points:

@sliquister wrote:
> I'm not sure either how necessary the compatibility is (yes, it's a breaking change, but the alert only helps with the kind of breaking change that's fixed with a simple `-I +unix`, not with breakage of things like `ocamlopt $(ocamlc -where)/unix.cmxa`).

From testing on opam-repository (very much not exhaustive, obviously), there is literally one package which isn't already broken in OCaml 5.0 which does this, and that was an old version, however there are many packages which expect to be able to do `ocaml unix.cma` and more than one might expect which pass `unix.cmxa` directly to `ocamlopt`, even if only in a `configure` script.

@shindere wrote:
> I didn't read any comment saying we should take so much care to
backward compatibility, neither in temrs of alerts, nor in temrs
of installing.

I linked 36 pull requests in other repositories to this one, and those are just the ones I patched while investigating the impact of this change without any compatibility concern! For the installing, I linked #1724, #1728 and #10301, where we've done that before.

> So I'd avoid as much complexity as possible.

Side-stepping the fact that there was a typo in a variable (although Jenkins would have caught that eventually), the complexity we're talking about here is a few `rm` commands!

> My only comment about the code is regarding `"$(INSTALL_LIBDIR_LIBNAME)/"`.
Why insist onhaving the final `/`?

I hadn't noticed that I'd done that! I removed the one on L104 (it's now `$(MKDIR) "$(INSTALL_LIBDIR_LIBNAME)"`). For the others, I was leaving things as they were. For install, it's arguably better to have the trailing slash - `install foo bar/` errors if `bar` is accidentally a file (there's also install's `-t` flag).

> Would also love to have `/usr/lib/ocaml` to contain only subdirectories, and
having one for the runtime and one for the stdlib looks very sensible to me,
withthe C stubs in the
same directory than the library they belong to.

Indeed - I'll try to get a couple of further PRs opened with these, just for perusal!

> It feels like, once all this has been sorted out, it would be time to revisite #11007.

Yes, definitely - that can hopefully get some noise and be pushed over the proverbial line between now and branching.

dra27 commented:
I still need to check the places where `Load_path.init` doesn't invoke the hook and I'll try to schedule an opam-health-check run of a version of this with no compatibility/alert to see what the damage looks like.

Octachron commented on utils/warnings.mli:
```diff
@@ -26,6 +26,9 @@ type loc = {
 }
 
+val ghost_loc_in_file : string -> loc
+(** Return an empty ghost range range located in a given file *)
```
"range range" ⇒ range

dra27 replied:
Oops, fixed!

v-gb commented on utils/clflags.ml:
```diff
@@ -577,3 +577,26 @@ let create_usage_msg program =
+let auto_include_otherlibs alert uncap fn =
+  if !no_std_include then raise Not_found
+  else
+    let expand = Misc.expand_directory Config.standard_library in
+    let read_lib lib = lazy (Load_path.Dir.create (expand ("+" ^ lib))) in
+    let otherlibs =
+      List.map (fun lib -> (lib, read_lib lib)) ["dynlink"; "str"; "unix"] in
+    let scan (lib, dir) =
+      let dir = Lazy.force dir in
+      let check base =
+        if uncap base = fn then begin
+          alert lib;
+          Load_path.append_dir dir;
+          Some (Filename.concat (Load_path.Dir.path dir) base)
+        end else
+          None
+      in
+      List.find_map check (Load_path.Dir.files dir)
+    in
+    match List.find_map scan otherlibs with
+    | Some base -> base
+    | None -> raise Not_found
```
I think this works, although I think something like this would be clearer:

In `load_path.ml`
```ocaml
with Not_found ->
  (* call auto_include_callback like this instead, or the uncap version for the second call *)
  auto_include_callback (fun dir -> Misc.find_in_path [ dir ] fn)
```

here:
```ocaml
let auto_include alert find_in_dir =
  if !no_std_include then raise Not_found
  else
    List.find_map [ "dynlink"; "str"; "unix" ] ~f:(fun lib ->
      let dir = Filename.concat Config.standard_library lib in
      match find_in_dir dir with
      | Not_found -> None
      | x -> alert lib; Load_path.append_dir dir; Some x)
    (* then turn the option back into Not_found *)
```

I wonder if there's a cost to checking for file existence for every missing file. It might seem like a missing file will result in a compilation failure immediately but that's not the case: the compiler can look for cmi or cmx files that may well be missing and proceed. Not sure if that's a real concern, to be clear, but that's why I was suggesting before to special case the cmis that are known to live in these libs: it means no extra filesystem interaction.

dra27 replied:
I agree - as you say, contrary to expectation, it certainly does matter not to be scanning these directories more than once. Both versions of the code I produced have aimed to ensure that those directories are never scanned additionally if you have specified the correct `-I +dir` (i.e. we in no way penalize calls which are correct) and that each directory is only scanned once, if at all. However, I messed up the laziness in this revised version, which I've just fixed. I'm having a think about whether there's a neat way to pass the search function _and_ keep the laziness...

dra27 replied:
I just pushed an amended version - instead of using `Misc.find_in_path`, I added `Load_path.Dir.find` which allows the two call-sites in `Load_path` to pass either `Load_path.Dir.find` or `Load_path.Dir.find_uncap`. I agree that the result looks much clearer - and it still gets to keep the lazy scanning as well (because the `Dir.t` is generated lazily).

v-gb commented on parsing/location.ml:
```diff
@@ -897,6 +895,15 @@ let alert ?(def = none) ?(use = none) ~kind loc message =
 
+let auto_include_alert lib =
+  let message = Printf.sprintf "\
+    OCaml's lib directory layout changed in 5.0. The %s subdirectory has been \
+    automatically added to the search path, but you should add -I +%s to the \
+    command-line to silence this alert." lib lib in
```
I suspect most project use dune these days, maybe this is slightly better advice?

```suggestion
    OCaml's lib directory layout changed in 5.0. The %s subdirectory has been \
    automatically added to the search path, but you should add -I +%s to the \
    command-line (or add %s to the list of libraries in the dune file) to silence \
    this alert." lib lib lib in
```

dra27 replied:
Hmm - I like the idea, but I'm loathe to make it completely Dune-specific. How about

```suggestion
    OCaml's lib directory layout changed in 5.0. The %s subdirectory has been \
    automatically added to the search path, but you should add -I +%s to the \
    command-line to silence this alert (e.g. by adding %s to the list of \
    libraries in your dune file, or adding use_%s to your _tags file for \
    ocamlbuild, or using -package %s for ocamlfind)." lib lib lib lib lib in
```

v-gb replied:
Sure!

Octachron replied:
In term of formatting, since the alert reporting is using the message string directly, the line breaks need be added directly to the message. One option would beto add  `\n` line break directly, the other would be to switch to `Format.pp_print_text` with 

```ocaml
Format.asprintf "%a" Format.pp_print_text  
  "OCaml's lib directory layout changed in 5.0...."
```

Octachron replied:
In term of formatting, since the alert reporting is using the message string directly, the line breaks need to be added explicitly to the message. One option would be to add  `\n` line break directly, the other would be to switch to `Format.pp_print_text` with 

```ocaml
Format.asprintf "%a" Format.pp_print_text  
  "OCaml's lib directory layout changed in 5.0...."
```

shindere commented:
Out of curiosity: in first commit, why do you use `?=` when defining
`INSTALL_LIBDIR_LIBNAME` ? I feel I would have done quite the countrary.

Apart from that, the code looks good to me so I am approving the PR,
but would rather leave it to other developers to decide whether
this is the right way to go.

dra27 commented:
> Out of curiosity: in first commit, why do you use `?=` when defining `INSTALL_LIBDIR_LIBNAME` ? I feel I would have done quite the countrary.

Ah, this was an artefact from testing on 4.14, where this allowed the now-removed `otherlibs/bigarray/Makefile` to override it and still install to the "old" location. I've changed it to `=`

shindere commented:
David Allsopp (2022/05/10 02:44 -0700):
> Ah, this was an artefact from testing on 4.14, where this allowed the
> now-removed `otherlibs/bigarray/Makefile` to override it and still
> install to the "old" location. I've changed it to `=`

Cool! Thanks!

Octachron commented on parsing/location.ml:
```diff
@@ -897,6 +895,15 @@ let alert ?(def = none) ?(use = none) ~kind loc message =
+    automatically added to the search path, but you should add -I +%s to the \
+    command-line to silence this alert." lib lib in
+  let alert =
+    {Warnings.kind="ocaml_auto_include"; use=none; def=none; message} in
```
I am wondering if we want to name the alert `ocaml_deprecated_auto_include` to indicate that the alert may be replaced by an error in some future.

dra27 replied:
Done - I've only changed the string, I assume the name of the function itself is OK without `_deprecated_` in it?

Octachron replied:
Yes, that's seem fine to me.

Octachron commented on utils/clflags.mli:
```diff
@@ -272,3 +272,10 @@ val print_arguments : string -> unit
 (* [reset_arguments ()] clear all declared arguments *)
 val reset_arguments : unit -> unit
+
+val auto_include_otherlibs :
```
I am not sure that I follow the reason why this function is in `Clflags`? My mental model is that `Clflags` does not interpret argument itself?

dra27 replied:
Is there somewhere slightly higher up where it can live in ocamlcommon?

dra27 replied:
The problem is that `Compmisc` would be an obvious place, except that `Ast_mapper` uses it...

Octachron replied:
What about moving the function to `Load_path` with a reference argument?

shindere commented:
Florian Angeletti (2022/05/10 04:59 -0700):
> I am wondering if we want to name the alert
> `ocaml_deprecated_auto_include` to indicate that the alert may be
> replaced by an error in some future.

Good idea! The turning into an error could even be explitly announced.

v-gb commented on utils/clflags.ml:
```diff
@@ -577,3 +577,24 @@ let create_usage_msg program =
+let auto_include_otherlibs otherlibs alert find_in_dir fn =
+  if !no_std_include then raise Not_found
+  else
+    let scan (lib, dir) =
+      let dir = Lazy.force dir in
```
```suggestion
    let scan (lib, lazy dir) =
```

dra27 replied:
I wish I could say I knew we could do that 🙈

v-gb commented on utils/load_path.mli:
```diff
@@ -31,7 +31,30 @@ val remove_dir : string -> unit
+      sub-directories of this directory. *)
+
+  val find : t -> string -> string option
+  (** [find dir fn] returns the full path to [fn] in [dir].
+      @raise Not_found if [fn] does not exist in [dir]. *)
```
```suggestion
  (** [find dir fn] returns the full path to [fn] in [dir]. *)
```

dra27 replied:
Oops!

v-gb reviewed (commented):
I observe some kind of flushing issue in the toplevel (on linux, if that makes a difference):

```
# Unix.openfile;;
# Unix.openfile;;include: OCaml's lib directory layout changed in 5.0. The unix subdirectory has been automatically added to the search path, but you should add -I +unix to the command-line to silence this alert.
Error: Reference to undefined global `Unix'
```

Independently, I think it'd be easier to pass in the extra argument to `Load_path.init` in debugger/program_management.ml and debugger/command_line.ml than to think about whether it is necessary or not.

But otherwise, looks good to me. It's certainly much simpler now!

dra27 commented:
> I observe some kind of flushing issue in the toplevel (on linux, if that makes a difference):
> 
> ```
> # Unix.openfile;;
> # Unix.openfile;;include: OCaml's lib directory layout changed in 5.0. The unix subdirectory has been automatically added to the search path, but you should add -I +unix to the command-line to silence this alert.
> Error: Reference to undefined global `Unix'
> ```

I observe this too, although not with `type t = Unix.file_descr;;`, for example. @Octachron - any idea what's going with the alert mechanism here not printing a new line first??

> Independently, I think it'd be easier to pass in the extra argument to `Load_path.init` in debugger/program_management.ml and debugger/command_line.ml than to think about whether it is necessary or not.

I've added them all in - especially now that it's a much smaller change to `Load_path.init`, my previous reticence to add them all makes no sense.

> But otherwise, looks good to me. It's certainly much simpler now!

Thanks!

v-gb commented:
> > I observe some kind of flushing issue in the toplevel (on linux, if that makes a difference):
> > ```
> > # Unix.openfile;;
> > # Unix.openfile;;include: OCaml's lib directory layout changed in 5.0. The unix subdirectory has been automatically added to the search path, but you should add -I +unix to the command-line to silence this alert.
> > Error: Reference to undefined global `Unix'
> > ```
> 
> I observe this too, although not with `type t = Unix.file_descr;;`, for example. @Octachron - any idea what's going with the alert mechanism here not printing a new line first??

Ah, the problem goes away if I reduce the size of my font so the alert fits on one line. The problem is that this highlighting code in location.ml counts the number of '\n' to count the number of line, but that doesn't work when lines wrap around. So let's just add some newlines in the text of the alert.

dra27 commented:
The alert message should now display correctly - by repeating a small amount of the code in `Ast_mapper`, the function now lives just about naturally in `Load_path` where it belongs with the main wrapper to it in `Compmisc`.

v-gb commented on utils/load_path.mli:
```diff
@@ -31,9 +31,37 @@ val remove_dir : string -> unit
+  (** All the files in that directory. This doesn't include files in
+      sub-directories of this directory. *)
+
+  val find : t -> string -> string option
+  (** [find dir fn] returns the full path to [fn] in [dir]. *)
+
+  val find_uncap : t -> string -> string option
+  (** As {!find}, but search also for uncapitalized name, i.e. if name is
+      Foo.ml, either /path/Foo.ml or /path/foo.ml may be returned. *)
```
These probably don't need to be exposed anymore?

dra27 replied:
They don't do any harm - the module itself is used elsewhere in the codebase, so it doesn't seem necessary to hide them.

v-gb commented on parsing/location.ml:
```diff
@@ -897,6 +895,19 @@ let alert ?(def = none) ?(use = none) ~kind loc message =
+    ocamlbuild, or using -package %s for ocamlfind)." lib lib lib lib lib in
+  let alert =
+    {Warnings.kind="ocaml_deprecated_auto_include"; use=none; def=none;
+     message = Format.asprintf "@[\n%a@]" Format.pp_print_text message}
```
Maybe it should be, so the format code understands the line break should be indented if necessary?

```suggestion
     message = Format.asprintf "@[@\n%a@]" Format.pp_print_text message}
```

dra27 replied:
I think it's better to have it being an explicit line break `@\n`, yes.

v-gb reviewed (approved):
I left a couple of comments, but this looks great.

Octachron commented on driver/compmisc.ml:
```diff
@@ -13,6 +13,10 @@
 
+let auto_include =
+  if !Clflags.no_std_include then fun _ _ -> raise Not_found
+  else Load_path.auto_include_otherlibs Location.auto_include_alert
```
This reads the value of `Clflags.no_std_include` when `Compmisc` is initialized which happens before CLI argument are parsed.

dra27 replied:
Oops (again) - done!

v-gb commented on parsing/ast_mapper.ml:
```diff
@@ -903,7 +903,11 @@ module PpxContext = struct
+            if !Clflags.no_std_include then fun _ _ -> raise Not_found
+            else Load_path.auto_include_otherlibs Location.auto_include_alert
+          in
+          Load_path.init ~auto_include (get_list get_string payload)
```
Maybe the same change as in compmisc should be made here. I don't think it will change the behavior in this case, but it's more clearly consistent to do the same thing.

dra27 replied:
Yeah, the consistency is definitely better - in fact, I added a comment to the recent push stating that it duplicates `Compmisc.auto_include`

Octachron commented:
The current implementation of the deprecation alert seems quite light and should be easy enough to remove in future versions of OCaml. I think that it offers quite a good compromise in term of usability: we are maintaining backward compatibility, while offering a migration path to the more future compatible setup.

dra27 commented:
Thanks for the approval, @Octachron! I’d like to hold the merging to allow synchronisation with the Dune and ocamlfind PRs so that ecosystem testing doesn’t have to pause.

dra27 commented:
Thank you everyone for the reviewing time and feedback on this - it's a nice change to have got over the line for 5.0!

gasche commented:
I just observed the following interaction in the OCaml toplevel which I think is a bit confusing for non-experts.

```ocaml
$ ocaml
OCaml version 5.0.0
Enter #help;; for help.

# Unix.sleep 1;;
Alert ocaml_deprecated_auto_include: 
OCaml's lib directory layout changed in 5.0. The unix subdirectory has been
automatically added to the search path, but you should add -I +unix to the
command-line to silence this alert (e.g. by adding unix to the list of
libraries in your dune file, or adding use_unix to your _tags file for
ocamlbuild, or using -package unix for ocamlfind).
Error: Reference to undefined global `Unix'
```

Is there something easy we can do to give better advice here?

Note: I you start with `#load "unix.cma";;` you get this alert as well, the "right" way to do things in the toplevel now is `#directory "+unix"` then `#load "unix.cma"`.

Octachron commented:
The easy way in the toplevel is rather `#require "unix";;`.  We could tune the alert message in the toplevel to point in this direction.

dra27 commented:
I'm not sure we should have the compiler recommend a command which requires findlib, though? In this particular case, I think what might be wanted is _not_ to display the alert?

dra27 commented:
I've not double-checked, but I think we'd be able to defer displaying the alert for long enough in the toplevel that we'd have the undefined global case already. The alert does want to be displayed if you do something like `type t = Unix.file_descr`, of course.

Octachron commented:
I think that having an alert in itself is fine? The toplevel user that uses low-level (aka not-library-aware) commands ought to first add the `unix` directory to its path and then load the library `cmxa`.

dra27 commented:
My thinking is that the code is already an error in 4.x, the alert is meant to warn that accidental use of the _types_ might become an error in future?

dra27 commented:
i.e. that the alert is displayed only in cases where OCaml 4.x would not have displayed anything, because it silently worked (because `unix.cmi` was in the stdlib directory)

Octachron commented:
`Unix.sleep` failing on an undefined global rather than an unbound identifier is in itself an observable effect of the special casing of `Unix`.

I think that having a strange and unexplained difference between `Mylib.f` and `Unix.f` is already worth an explanation rather than trying to hide what is happening because the code is ultimately failing.

dra27 commented:
Oops, yes, I follow your point now. It feels bad to me for us to recommend a directive when it isn't present (I'm sure there are still plenty of first-time users who run the `ocaml` command before discovering either `utop` or `#use "topfind"`). However, the toplevel does _know_ the directives which have been added - what you would both think to the alert recommending `#require` if it detects that a `"require"` directive has been added? It's mildly heretical, but in practice `topfind` and `utop` mean that `#require` in effect has only one meaning? Otherwise, it should display the `#directory` and `#use` runes.

dbuenzli commented:
As we see sometimes trying to be smart makes things more confusing. Why not simply let it go without crazy special casing ? There's release notes to document these changes.

nojb commented:
> As we see sometimes trying to be smart makes things more confusing.

Just my 2c: personally I am also of the opinion that the "automatic" include logic is not worth the hassle.

dra27 commented:
That's be the target, yes - but I'm not sure we can get away with completing a deprecation made in 5.0 in 5.1? I was intending - as with all new deprecations - to check when 5.1 is branched how many packages would still fail if this were made a hard error (and do another round of PRs, like the list above).

gasche commented:
My thinking:
- I agree with @dbuenzli that we don't want to add more complexity here.
- What is the behavior when using `utop` or `down/topfind`? If users there are not affected, maybe doing nothing is an option. (There are other ways raw `ocaml` is unpleasant.)

dra27 commented:
`utop` has the correct behaviour (`Unix` is not a special case in any way):

```
utop # Unix.sleep 1;;
Error: Unbound module Unix
Hint: Did you mean Unit?
```

It was able to do that for 5.0 because it's not used for executing scripts in the same way as `ocaml`. We could for 5.1 remove the automatic inclusion completely for `ocaml` as well (and then do that for the compilers in a further release or so's time)?

gasche commented:
> We could for 5.1 remove the automatic inclusion completely for ocaml as well (and then do that for the compilers in a further release or so's time)?

This sounds like a good idea to me.
